shell/platform/darwin/common/framework/Headers/FlutterMacros.h

```diff
@@ -37,4 +37,10 @@
  */
 #define FLUTTER_UNAVAILABLE(msg) __attribute__((__unavailable__(msg)))
 
+#if __has_feature(objc_arc)
+#define FLUTTER_ASSERT_ARC
+#else
+#define FLUTTER_ASSERT_ARC #error ARC must be enabled!
```

Comment:
This doesn't do what you want. It will abort execution but due to the preprocessor substitution not resulting in an expression (instead, another preprocessor statement). So the nice error message will be replaced by something like the preprocessor barfing with an error that says a "statement" was necessary (but wont actually tell you the statement just the source line).

I suggest just staying away from the preprocessor for the second bit. Something like:

`static_assert(__has_feature(objc_arc), "This translation unit should be compiled with ARC")`.

Reply:
Let's also add a `FLUTTER_ASSERT_NOT_ARC` or something similar. Now that we are mixing arc and non-arc TUs in the repo, it will be easier to scroll up to the file and check if the review needs to assume arc or not.

Reply:
`FLUTTER_ASSERT_NOT_ARC` would be false when the `FlutterMacros.h` isn't imported, which might be easy to forget to do.

Reply:
An example of the error https://godbolt.org/z/JpJdWu

Reply:
> `FLUTTER_ASSERT_NOT_ARC` would be false when the `FlutterMacros.h` isn't imported, which might be easy to forget to do.

Can OBJC_ARC_UNAVAILABLE be used instead of FLUTTER_ASSERT_NOT_ARC?

Reply:
> static_assert(__has_feature(objc_arc), "This translation unit should be compiled with ARC")

`static_assert` is a c++ feature so we won't always have access to it.

> `FLUTTER_ASSERT_NOT_ARC` would be false when the `FlutterMacros.h` isn't imported, which might be easy to forget to do.

Undefined macros should be a compilation error, we can fiddle our flags if that isn't already the case.

> Can OBJC_ARC_UNAVAILABLE be used instead of FLUTTER_ASSERT_NOT_ARC?

I don't believe so.  That is used for hiding methods from ARC compiled files.

Reply:
> An example of the error https://godbolt.org/z/JpJdWu

Clang will give the line number and people can just jump to that and see the #error message.

Reply:
`_Static_assert` is C11 and available in C++.

Reply:
> > static_assert(__has_feature(objc_arc), "This translation unit should be compiled with ARC")
> 
> `static_assert` is a c++ feature so we won't always have access to it.
> 
> > `FLUTTER_ASSERT_NOT_ARC` would be false when the `FlutterMacros.h` isn't imported, which might be easy to forget to do.
> 
> Undefined macros should be a compilation error, we can fiddle our flags if that isn't already the case.
> 
> > Can OBJC_ARC_UNAVAILABLE be used instead of FLUTTER_ASSERT_NOT_ARC?
> 
> I don't believe so. That is used for hiding methods from ARC compiled files.

I guess I don't know how this would be used.  There are things _within_ the class/method that are not available in ARC that need to be pre-processed out?  So we can't follow NSObject's patterns:
```objc
- (instancetype)retain OBJC_ARC_UNAVAILABLE;
- (oneway void)release OBJC_ARC_UNAVAILABLE;
- (instancetype)autorelease OBJC_ARC_UNAVAILABLE;
- (NSUInteger)retainCount OBJC_ARC_UNAVAILABLE;
- (struct _NSZone *)zone OBJC_ARC_UNAVAILABLE;
```

Reply:
I'll defer to @chinmaygarde here.  Whatever you prefer is fine.

Reply:
@chinmaygarde it looks like the preprocessor error gives a meaningful error, whilst _Static_assert doesn't:

_Static_assert:
<img width="722" alt="Screen Shot 2019-09-30 at 2 35 22 PM" src="https://user-images.githubusercontent.com/30870216/65918899-bc455680-e38f-11e9-8dff-d534ac4aeff7.png">

#error:
<img width="716" alt="Screen Shot 2019-09-30 at 2 35 48 PM" src="https://user-images.githubusercontent.com/30870216/65918919-c8c9af00-e38f-11e9-91bf-71da02a0cea2.png">